sdk/core/System.ClientModel/src/Internal/ArrayBackedPropertyBag.cs

```diff
@@ -0,0 +1,293 @@
+                _first = (key, value);
+                _count = 1;
+                return;
+            case 1:
+                if (IsFirst(key))
+                {
+                    _first = (_first.Key, value);
+                }
+                else
+                {
+                    _second = (key, value);
+                    _count = 2;
+                }
+
+                return;
+            default:
+                if (_rest == null)
+                {
+                    _rest = ArrayPool<(TKey Key, TValue Value)>.Shared.Rent(8);
+                    _rest[_count++ - 2] = (key, value);
+                    return;
+                }
+
+                if (_rest.Length <= _count)
+                {
+                    var larger = ArrayPool<(TKey Key, TValue Value)>.Shared.Rent(_rest.Length << 1);
+                    _rest.CopyTo(larger, 0);
+                    var old = _rest;
+                    _rest = larger;
+                    ArrayPool<(TKey Key, TValue Value)>.Shared.Return(old, true);
+                }
+                _rest[_count++ - 2] = (key, value);
+                return;
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void SetAt(int index, (TKey Key, TValue Value) value)
+    {
+        if (index == 0)
+            _first = value;
+        else if (index == 1)
+            _second = value;
+        else
+            GetRest()[index - 2] = value;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private TValue GetAt(int index) => index switch
+    {
+        0 => _first.Value,
+        1 => _second.Value,
+        _ => GetRest()[index - 2].Value
+    };
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private int GetIndex(TKey key)
+    {
+        if (_count == 0)
+            return -1;
+        if (_count > 0 && _first.Key.Equals(key))
+            return 0;
+        if (_count > 1 && _second.Key.Equals(key))
+            return 1;
+
+        if (_count <= 2)
+            return -1;
+
+        (TKey Key, TValue Value)[] rest = GetRest();
+        int max = _count - 2;
+        for (var i = 0; i < max; i++)
+        {
+            if (rest[i].Key.Equals(key))
+                return i + 2;
+        }
+        return -1;
+    }
+
+    public void Dispose()
+    {
+#if DEBUG
+        if (_disposed)
+        {
+            return;
+        }
+        _disposed = true;
+#endif
+        _count = 0;
+        _first = default;
+        _second = default;
+        if (_rest == default)
+        {
+            return;
+        }
+
+        lock (_lock)
+        {
+            var rest = _rest;
+            _rest = default;
+            ArrayPool<(TKey Key, TValue Value)>.Shared.Return(rest, true);
+        }
+    }
+
+    private (TKey Key, TValue Value)[] GetRest() => _rest ?? throw new InvalidOperationException($"{nameof(_rest)} field is null while {nameof(_count)} == {_count}");
+
+    private void CheckDisposed()
```

Comment:
Oh, why did you remove the conditional and not this ifdef?  The conditional attribute was making sure/clear that this method is not called in release builds. As it's written now, the method is called and then we have to rely on the inliner to make it be an no-op (free)

Reply:
I will put the attribute back.  I'd prefer to keep the ifdefs in place as they are consistent with other methods.  I realize this is redundant but unless there's a clear problem with that redundancy, I prefer it.